proto/schema.proto

```diff
@@ -0,0 +1,170 @@
+syntax="proto3";
+
+option go_package = "github.com/kubernetes-csi/external-snapshot-metadata/pkg/grpc";
+
+service SnapshotMetadata {
+  rpc GetMetadataAllocated(GetMetadataAllocatedRequest)
+    returns (stream GetMetadataAllocatedResponse) {}
+
+  rpc GetMetadataDelta(GetMetadataDeltaRequest)
+    returns (stream GetMetadataDeltaResponse) {}
+}
+
+enum BlockMetadataType {
+  UNKNOWN=0;
+
+  // The FIXED_LENGTH value indicates that data ranges are
+  // returned in fixed size blocks.
+  FIXED_LENGTH=1;
+
+  // The VARIABLE_LENGTH value indicates that data ranges
+  // are returned in potentially variable sized extents.
+  VARIABLE_LENGTH=2;
+}
+
+// BlockMetadata specifies a data range.
+message BlockMetadata {
+  // This is the zero based byte position in the volume or snapshot,
+  // measured from the start of the object.
+  // This field is REQUIRED.
+  uint64 byte_offset = 1;
+
+  // This is the size of the data range.
+  // size_bytes MUST be greater than zero.
+  // This field is REQUIRED.
+  uint64 size_bytes = 2;
+}
+
+// The GetMetadataAllocatedRequest message is used to solicit metadata
+// on the allocated blocks of a snapshot: i.e. this identifies the
+// data ranges that have valid data as they were the target of some
+// previous write operation on the volume.
+message GetMetadataAllocatedRequest {
+  // This field specifies an audience scoped authentication token presented by client.
+  // This token is used to authenticate and authorize the request at server side.
+  // This field is REQUIRED.
+  string security_token = 1;
+
+  // This field indicates the Kubernetes namespace where the VolumeSnapshot exists.
+  // If not set, the VolumeSnapshot will be looked for in the "default" namespace.
+  // This field is OPTIONAL.
+  string namespace = 2;
+
+  // This is the VolumeSnapshot object name against which changes
+  // are to be computed.
+  // This field is REQUIRED.
+  string snapshot_name = 3;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed.
+  // It is intended to be used to continue a previously interrupted
+  // call.
+  // The client SHOULD specify this value to be the offset of the byte
+  // position immediately after the last byte of the last data range
+  // received, if continuing an interrupted operation, or zero if not.
```

Comment:
The rest of the comment in the CSI spec provides a guarantee to the client.  Translated to this domain it would be:

>  // The sidecar will ensure that the returned response stream does not
>  // contain BlockMetadata tuples that end before the requested
>  // starting_offset: i.e. if S is the requested starting_offset, and
>  // B0 is block_metadata[0] of the first message in the response
>  // stream, then (S < B0.byte_offset + B0.size_bytes) will be true.